modules/phase_field/src/postprocessors/FeatureFloodCount.C

```diff
@@ -727,56 +679,57 @@ FeatureFloodCount::visitElementalNeighbors(const Elem * elem, unsigned long curr
-       * entities may or may not end up being part of the feature.
-       * We will not update the _entities_visited data structure
-       * here.
-       */
-      feature->_halo_ids.insert(neighbor_ptr->id());
-
-      if (recurse)
-        flood(neighbor_ptr, current_idx, feature);
-    }
-  }
+  visitNeighborsHelper(elem, all_active_neighbors, current_idx, feature, expand_halos_only);
 }
 
 void
-FeatureFloodCount::visitNodalNeighbors(const Node * node, unsigned long current_idx, FeatureData * feature, bool recurse)
+FeatureFloodCount::visitNodalNeighbors(const Node * node, unsigned long current_idx, FeatureData * feature, bool expand_halos_only)
 {
   mooseAssert(node, "Node is NULL");
 
-  std::vector<const Node *> neighbors;
-  MeshTools::find_nodal_neighbors(_mesh.getMesh(), *node, _nodes_to_elem_map, neighbors);
+  std::vector<const Node *> all_active_neighbors;
+  MeshTools::find_nodal_neighbors(_mesh.getMesh(), *node, _nodes_to_elem_map, all_active_neighbors);
 
-  // Loop over all nodal neighbors
-  for (unsigned int i = 0; i < neighbors.size(); ++i)
-  {
-    const Node * neighbor_node = neighbors[i];
-    auto my_proc_id = processor_id();
+  visitNeighborsHelper(node, all_active_neighbors, current_idx, feature, expand_halos_only);
+}
 
-    // Only recurse on nodes this processor can see
-    if (_mesh.isSemiLocal(const_cast<Node *>(neighbor_node)))
+template<typename T>
+void
+FeatureFloodCount::visitNeighborsHelper(const T * curr_entity, std::vector<const T *> neighbor_entities, unsigned long current_idx,
+                                        FeatureData * feature, bool expand_halos_only)
+{
+  // Loop over all active element neighbors
+  for (const auto neighbor : neighbor_entities)
+  {
+    if (neighbor)
     {
-      if (neighbor_node->processor_id() != my_proc_id)
-        feature->_ghosted_ids.insert(neighbor_node->id());
+      if (expand_halos_only)
+        feature->_halo_ids.insert(neighbor->id());
+
+      else
+      {
+        auto my_processor_id = processor_id();
 
-      // Premark Halo values
-      feature->_halo_ids.insert(neighbor_node->id());
+        if (neighbor->processor_id() != my_processor_id)
+          feature->_ghosted_ids.insert(curr_entity->id());
 
-      if (recurse)
-        flood(neighbors[i], current_idx, feature);
+        /**
+         * Only recurse where we own this entity. We might step outside of the
+         * ghosted region if we recurse where we don't own the current entity.
+         */
+        if (curr_entity->processor_id() == my_processor_id)
+        {
+          /**
+           * Premark neighboring entities with a halo mark. These
+           * entities may or may not end up being part of the feature.
+           * We will not update the _entities_visited data structure
+           * here.
+           */
+          feature->_halo_ids.insert(neighbor->id());
+
+          flood(neighbor, current_idx, feature);
+        }
```

Comment:
@roystgnr - Take a look at this condition for recursing. It's actually quite simple now and much more elegant that looking at the semi-local set. Also seems to work just fine :)